### PR TITLE
support dropout in pytorch training

### PR DESCRIPTION
**Pull request by fanlu**
- dropout with dropout_scheduler `0,0@0.20,0.5@0.50,0` will give 0.1-0.2 percentage better
- `dataloader.sampler.set_epoch(epoch)` make the result a little better
- change Linear to conv1d in tdnn module to avoid  two permute operation

### Result
***
||tdnn_1c_rd_rmc_rng_without_cleanup | tdnn_1c_rd_rmc_rng|
|--|-- | --|
|dev_cer | 5.92 | 5.99|
|dev_wer | 13.71 | 13.86|
|test_cer | 7.03 | 7.08|
|test_wer | 15.35|15.72|

||TDNN-F(Pytorch, Adam, dropout + ivector ) | TDNN-F(Pytorch, Adam, dropout  without ivector ) |
|--|-- | --|
|dev_cer | 6.00 | 6.13|
|dev_wer | 13.72 | 13.97|
|test_cer | 7.20 | 7.19|
|test_wer | 15.52|15.55|
|global objf|-0.046589|-0.050291|
|valid objf|-0.053605|-0.062898|
|output-affine-norm|76.83|83.2|

@csukuangfj @qindazhu @danpovey please have a reivew.

### CODE REVIEW

**Comment by danpovey** [comment]
Thanks!!  I had a quick look.  If I hear from someone else that it's OK to merge, I'll merge.

File: egs/aishell/s10/chain/tdnnf_layer.py [hunk 1]
```diff
@@ -137,6 +159,42 @@ def forward(self, x):
...
+        
+        self.dropout = SharedDimScaleDropout(dim=2)
+
+    def forward(self, x, dropout=0.):
+        # input x is of shape: [batch_size, feat_dim, seq_len] = [N, C, T]
+        x = self.affine(x)
+        # at this point, x is [N, C, T]
+        
+        x = F.relu(x)
+        # at this point, x is [N, C, T]
```
Comment (qindazhu) [inline egs/aishell/s10/chain/tdnnf_layer.py]: I guess we can delete these comments now as there's no `permute` operation and the shape of `x` is clear.

Comment (fanlu) [reply egs/aishell/s10/chain/tdnnf_layer.py]: OK

File: egs/aishell/s10/chain/tdnnf_layer.py [hunk 2]
```diff
@@ -137,6 +159,35 @@ def forward(self, x):
...
+                                              affine=False)
+        
+        self.dropout = SharedDimScaleDropout(dim=2)
+
+    def forward(self, x, dropout=0.):
+        # input x is of shape: [batch_size, feat_dim, seq_len] = [N, C, T]
+        x = self.affine(x)
+        x = F.relu(x)
+        x = self.batchnorm(x)
+        x = self.dropout(x, alpha=dropout)
```
Comment (csukuangfj) [inline egs/aishell/s10/chain/tdnnf_layer.py]: Add a guard here:

```python
if self.training:
  x = self.dropout(x, alpha=dropout)
```

The same place for the following `FactorizedTDNN`.

File: egs/aishell/s10/local/run_chain.sh [hunk 3]
```diff
@@ -67,7 +69,7 @@ if [[ $stage -le 3 ]]; then
 fi
 
 train_ivector_dir=
-if [[ $train_ivector ]]; then
+if [ "$train_ivector" = true ]; then
```
Comment (csukuangfj) [inline egs/aishell/s10/local/run_chain.sh]: If `train_ivector` is going to have only two values: `true` or `false`, then
you can use

```bash
if $train_ivector; then
fi
```

File: egs/aishell/s10/local/run_chain.sh [hunk 4]
```diff
@@ -255,6 +263,7 @@ if [[ $stage -le 17 ]]; then
         --train.ddp.init-method $init_method \
         --train.ddp.multiple-machine $use_multiple_machine \
         --train.ddp.world-size $world_size \
+        --train.dropout-schedule "$dropout_schedule" \
         --train.cegs-dir $dir/merged_egs \
```
Comment (csukuangfj) [inline egs/aishell/s10/local/run_chain.sh]: please sort the options alphabetically.

File: egs/aishell/s10/chain/tdnnf_layer.py [hunk 5]
```diff
@@ -51,6 +51,28 @@ def _constrain_orthonormal_internal(M):
     return M
 
 
+class SharedDimScaleDropout(nn.Module):
+    def __init__(self, dim=1):
+        '''
+        Continuous scaled dropout that is const over chosen dim (usually across time)
+        Multiplies inputs by random mask taken from Uniform([1 - 2\alpha, 1 + 2\alpha])
+        '''
+        super(SharedDimScaleDropout, self).__init__()
```
Comment (csukuangfj) [inline egs/aishell/s10/chain/tdnnf_layer.py]: Use
```python
super().__init__()
```

File: egs/aishell/s10/chain/tdnnf_layer.py (hunk 2 above)
Comment (fanlu) [reply egs/aishell/s10/chain/tdnnf_layer.py]: self.training is in dropout module forward function,
```
if self.training and alpha > 0.:
```
Should I change it?

Comment (csukuangfj) [reply egs/aishell/s10/chain/tdnnf_layer.py]: I see. Just keep it unchanged.

**Comment by fanlu** [comment]
@danpovey please merge this pr, then I can open another one about lda.

**Comment by csukuangfj** [comment]
+1
